### PR TITLE
Implement Dora client side affinity

### PR DESCRIPTION
**Pull request by beinan**
### What changes are proposed in this pull request?

Dora client side affinity

### Why are the changes needed?

skip grpc call of the get block location from master

### Does this PR introduce any user facing changes?
new configruations

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+  @Override
+  public void seek(long pos) throws IOException {
+    Preconditions.checkState(!mClosed, "Cannot do operations on a closed BlockInStream");
+    Preconditions.checkArgument(pos >= 0, PreconditionMessage.ERR_SEEK_NEGATIVE.toString(), pos);
+    Preconditions.checkArgument(pos <= mLength,
+        "Seek position past the end of the read region (block or file).");
+    if (pos == mPos) {
+      return;
+    }
+    closeDataReader();
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java:146]: if the new position is still within the range of the current chunk, we do not have to close the reader?

File: core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  }
+
+  @Override
+  public long skip(long n) throws IOException {
+    Preconditions.checkState(!mClosed, "Cannot do operations on a closed BlockInStream");
+    if (n <= 0) {
+      return 0;
+    }
+    long toSkip = Math.min(remaining(), n);
+    mPos += toSkip;
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java:157]: call `seek(mPos + toSkip)` instead?

File: core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheClient.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  /**
+   * Get status.
+   * @param path
+   * @param options
+   * @return uri status
+   */
+  public URIStatus getStatus(String path, GetStatusPOptions options) {
+    try {
+      WorkerNetAddress workerNetAddress = getWorkerNetAddress(path);
+      CloseableResource<BlockWorkerClient> client =
```
Comment (dbw9580) [inline core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheClient.java:84]: Wrap in try with resource:

```suggestion
      try (CloseableResource<BlockWorkerClient> client =
          mContext.acquireBlockWorkerClient(workerNetAddress)) {
```

File: tests/src/test/java/alluxio/client/fs/io/DoraReadIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import com.google.common.io.ByteStreams;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.io.InputStream;
+
+public class DoraReadIntegrationTest extends BaseIntegrationTest {
+  private static final int TEST_FILE_LENGTH = 50 * Constants.KB;
```
Comment (dbw9580) [inline tests/src/test/java/alluxio/client/fs/io/DoraReadIntegrationTest.java:40]: More test file lengths to cover the corner cases?

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    try (DataReader reader = mGrpcReaderFactory.create(position, length)) {
+      while (totalBytesRead < length) {
+        DataBuffer dataBuffer = null;
+        try {
+          dataBuffer = reader.readChunk();
+          if (dataBuffer == null) {
+            break;
+          }
+          int bytesRead = dataBuffer.readableBytes();
+          dataBuffer.readBytes(buffer, offset, bytesRead);
```
[the remark this answers is not in the document]
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java:116]: good call,  fixed

File: core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java (hunk 1 above)
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java:146]: hmmm, very likely you're right, but I just follow the code in the blockInStream.  I'm thinking if we could optimize this in another pr?

File: core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java (hunk 2 above)
Comment (beinan) [reply core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java:157]: done

File: tests/src/test/java/alluxio/client/fs/io/DoraReadIntegrationTest.java (hunk 4 above)
Comment (beinan) [reply tests/src/test/java/alluxio/client/fs/io/DoraReadIntegrationTest.java]: done

**Comment by JySongWithZhangCe** [comment]
Affinity scheduling is a very good feature, which can effectively alleviate the current memory imbalance problem of Alluxio. When the computing engines use the same hash algorithm, data locality will not be affected, which solves the current problem that Alluxio cannot guarantee data locality during load balancing. But at the same time, hotspot issues can cause data locality to fail. Is it possible for us to provide more than one cache address for hot data, for example, we detect hot data, and then use another hash algorithm to cache this part of data in the cluster.

**Comment by beinan** [comment]
> Affinity scheduling is a very good feature, which can effectively alleviate the current memory imbalance problem of Alluxio. When the computing engines use the same hash algorithm, data locality will not be affected, which solves the current problem that Alluxio cannot guarantee data locality during cache balancing. But at the same time, hotspot issues can cause data locality to fail. Is it possible for us to provide more than one cache replicate for hot data, for example, we detect hot data, and then use another hash algorithm to cache this part of data in the cluster.

Good call!  Thanks a lot for the comments!   Yes, we will support the replication very soon (that's why I made the hash functions return multiple nodes)

File: core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java (hunk 1 above)
Comment (dbw9580) [reply core/client/fs/src/main/java/alluxio/client/file/dora/DoraCacheFileInStream.java:149]: sure not a blocker for this PR

File: core/transport/src/main/proto/proto.lock [hunk 6]
```diff
@@ -3277,6 +3277,21 @@
               }
             ]
           },
+          {
+            "name": "S3SyntaxOptions",
+            "fields": [
```
Comment (dbw9580) [inline core/transport/src/main/proto/proto.lock:3282]: https://github.com/Alluxio/alluxio/pull/16651 added the missing proto.lock changes. We can cherry pick that PR to dora branch.

**Comment by beinan** [comment]
alluxio-bot, merge this please.

**Review by JiamingMai — COMMENTED** [review]
LGTM

**Pull request merged**